Upgrade to Hudi 0.9.0. Integrate with Hudi's Metadata based listing and Bootstrap Features

umehrot2 opened the pull request:
Upgrade to Hudi 0.9.0. Integrate with Hudi's Metadata based listing and Bootstrap Features

Test plan - (Please fill in how you tested your changes)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

n3nash commented:
@vkorukanti Could you please help review this ?

umehrot2 commented:
Can someone help review this ?

vinothchandar commented:
Hi all, could we get some eyes on this please?  @zhenxiao could you help out or connect us with someone who can help us. ?

dborkar commented:
@imjalpreet @ashishtadose  @beinan This is an important PR for the Presto + Hudi integration. can one of you take a look? 

@aweisberg FYI.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
+    public static boolean shouldUseRecordReaderFromInputFormat(Configuration configuration, Storage storage,
+                                                               Map<String, String> customSplitInfo)
+    {
+        if (customSplitInfo != null && customSplitInfo.containsKey(HiveUtil.CUSTOM_FILE_SPLIT_CLASS_KEY)) {
```
static import HiveUtil.CUSTOM_FILE_SPLIT_CLASS_KEY

umehrot2 replied:
Done.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
+            return Arrays.stream(inputFormat.getClass().getAnnotations())
+                    .map(Annotation::annotationType)
+                    .map(Class::getSimpleName)
+                    .anyMatch(name -> name.equals("UseRecordReaderFromInputFormat"));
```
make "UseRecordReaderFromInputFormat" as a constant?

umehrot2 replied:
Done.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
+                                              String tablePath)
+    {
+        // Set Hudi metadata based listing configurations to be used through input format
+        conf.set(HoodieMetadataConfig.METADATA_ENABLE_PROP, String.valueOf(isPreferMetadataToListHudiFiles(session)));
+        conf.set(HoodieMetadataConfig.METADATA_VALIDATE_PROP, String.valueOf(isHudiMetadataVerificationEnabled(session)));
```
static import HoodieMetadataConfig.METADATA_ENABLE_PROP and HoodieMetadataConfig.METADATA_VALIDATE_PROP

umehrot2 replied:
Done.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
+        boolean hasUseSplitsAnnotation = Arrays.stream(inputFormat.getClass().getAnnotations())
+                .map(Annotation::annotationType)
+                .map(Class::getSimpleName)
+                .anyMatch(name -> name.equals("UseFileSplitsFromInputFormat"));
```
Make UseFileSplitsFromInputFormat as a constant ?

umehrot2 replied:
Done.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:
```diff
@@ -0,0 +1,115 @@
+        public HiveFileInfo next() throws IOException
+        {
+            FileStatus fileStatus = hoodieBaseFileIterator.next().getFileStatus();
+            String[] name = new String[]{"localhost:" + DFSConfigKeys.DFS_DATANODE_DEFAULT_PORT};
```
static import DFSConfigKeys.DFS_DATANODE_DEFAULT_PORT

umehrot2 replied:
Done.

beinan reviewed (commented):
looks good to me, thank you for the contribution!  just a couple of coding style issues.

vinothchandar commented:
Thanks @dborkar and @beinan ! 

@umehrot2 should we also upgrade to 0.8.0 given it has already been out with some fixes around metadata table as well?

umehrot2 commented:
@beinan sorry for the delay here. I have addressed your comments, resolved conflicts and upgraded to the latest Hudi release i.e 0.9.0. Can you please take another look ? Also, it seems the workflow is not running and requires a maintainer to trigger it.

rubenssoto commented:
hi @umehrot2 

I would like to try metadata listing but, I tried to compile your branch and I had some errors:
```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (checkstyle) on project presto-hive: You have 2 Checkstyle violations. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :presto-hive
```

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:
```diff
@@ -0,0 +1,119 @@
+public final class HudiDirectoryLister
+        implements DirectoryLister
+{
+    private static final Logger LOGGER = Logger.get(HudiDirectoryLister.class);
```
I know we have mixed naming convention for logger,  but I think we're using lower case "log" for the instance of Logger.

beinan commented on presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:
```diff
@@ -0,0 +1,119 @@
+        public HiveFileInfo next() throws IOException
+        {
+            FileStatus fileStatus = hoodieBaseFileIterator.next().getFileStatus();
+            String[] name = new String[]{"localhost:" + DFS_DATANODE_DEFAULT_PORT};
```
Looks like you hardcoded "localhost" here,  is it really you intend to do?

beinan reviewed (commented):
Thank you @umehrot2 !  It looks good to me except the hardcoded "localhost",  could you help to provide more context about why you're using a hardcoded "localhost" in your code?  Thanks!

vinothchandar commented:
@arunthirupathi thanks for the feedback. I think that makes sense. We can upgrade and solve any bundling/packaging first. then land the changes. 

@umehrot2 please let me know what you think. cc @codope as well

codope commented:
> I went through the changes, looks like it has lot of breaking changes.
> 
> 1. Some caches are removed.
> 2. Previously two class types were supported for Hudi, now only one class.
> 
> It would be great, if it can be organized into smaller changes, so that they can be reviewed easily.

@arunthirupathi Thanks for the review. I am going to break it down into 3 PRs:
1. Just the version upgrade.
2. Hudi bootstrap related changes.
3. Integrate with Hudi's metadata listing.
Will ensure that the new PRs address the comments on this one.

Meanwhile, could you please review the upgrade PR #16881 ? The other changes depend on this upgrade.
